marlowe/src/Language/Marlowe/Core/V1/Semantics/Next.hs

```diff
@@ -0,0 +1,290 @@
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE RecordWildCards #-}
+
+module Language.Marlowe.Core.V1.Semantics.Next
+  ( AmbiguousIntervalProvided(..)
+  , ApplicableGeneralizedInputs(..)
+  , CanChoose(..)
+  , CanDeposit(..)
+  , CanNotify(..)
+  , CanReduce(..)
+  , CaseIndex(..)
+  , Indexed(..)
+  , IsMerkleizedContinuation(..)
+  , Next(..)
+  , next
+  , toCaseIndex
+  ) where
+
+import Control.Applicative (Alternative((<|>)), Applicative((<*>)), (<$>))
+import Data.Aeson (FromJSON(parseJSON), KeyValue((.=)), ToJSON(toJSON), Value(Object), object, (.:))
+import Data.Aeson.Types ()
+import Data.Bifunctor (Bifunctor(first))
+import Data.Coerce (coerce)
+import Data.Foldable (Foldable(fold))
+import Data.List.Index (indexed)
+import Data.Maybe (catMaybes)
+import Data.Monoid as Haskell (First(First, getFirst), Monoid(mempty), (<>))
+
+import Deriving.Aeson (Generic)
+import Language.Marlowe.Core.V1.Semantics
+  ( ReduceResult(ContractQuiescent, RRAmbiguousTimeIntervalError)
+  , evalObservation
+  , evalValue
+  , reduceContractUntilQuiescent
+  )
+import Language.Marlowe.Core.V1.Semantics.Types
+  ( AccountId
+  , Action(Choice, Deposit, Notify)
+  , Bound
+  , Case(..)
+  , ChoiceId
+  , Contract(Close, When)
+  , Environment
+  , Party
+  , State
+  , Token
+  )
+import Language.Marlowe.Pretty (Pretty(..))
+import PlutusTx.Prelude (Either(..), Integer, Maybe(..), uncurry, ($), (.))
+import qualified Prelude as Haskell
+
+data Next
+  = Next
+      { canReduce::CanReduce
+      , applicableGeneralizedInputs :: ApplicableGeneralizedInputs}
+    deriving stock (Haskell.Show,Haskell.Eq,Haskell.Ord,Generic)
+    deriving anyclass (Pretty)
+
+data AmbiguousIntervalProvided = AmbiguousIntervalProvided
+    deriving stock (Haskell.Show,Haskell.Eq,Haskell.Ord,Generic)
+    deriving anyclass (Pretty)
+
+-- | Index of an applicable input matching its derived action index in a When [Case Action]
+newtype CaseIndex = CaseIndex {unIndex ::Integer}
+    deriving stock (Haskell.Show,Haskell.Eq,Haskell.Ord,Generic)
+    deriving anyclass (Pretty)
+    deriving (FromJSON,ToJSON)
+
+data ApplicableGeneralizedInputs
+    = ApplicableGeneralizedInputs
+     { canNotifyMaybe :: Maybe (Indexed CanNotify)
+     , deposits    :: [Indexed CanDeposit]
+     , choices     :: [Indexed CanChoose]}
+  deriving stock (Haskell.Show,Haskell.Eq,Haskell.Ord,Generic)
+  deriving anyclass (Pretty)
+
+data Indexed a
+  = Indexed CaseIndex a
+    deriving stock (Haskell.Show,Haskell.Eq,Haskell.Ord,Generic)
+    deriving anyclass (Pretty)
+
+toCaseIndex :: Indexed a -> CaseIndex
+toCaseIndex (Indexed c _) = c
+
+instance Haskell.Semigroup ApplicableGeneralizedInputs where
+  a <> b  = ApplicableGeneralizedInputs
+              (getFirst $ (coerce . canNotifyMaybe $ a) <> (coerce . canNotifyMaybe $ b)) -- First Monoid
+              (deposits a   <> deposits b) -- (++) monoid
+              (choices a    <> choices b ) -- (++) monoid
```

Comment:
I really doubt the default implementation of Monoid for `CanChoose` does the right thing

Reply:
CanChoose or CanNotify ? , I think he does because It is tested ...  
it's ~ to the monoid (First Maybe) => 
`Just y <> Just x = Just y`
`Nothing <> Just x = Just x`
`Nothing <> Nothing = Nothing`